controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -451,38 +572,173 @@ func (pp *portPublisher) updateEndpoints(endpoints *corev1.Endpoints) {
 			}
 		}
 	}
+	pp.addresses = newAddressSet
 	pp.exists = true
+	pp.metrics.incUpdates()
+	pp.metrics.setPods(len(pp.addresses.Addresses))
+	pp.metrics.setExists(true)
+}
+
+func (pp *portPublisher) addEndpointSlice(slice *discovery.EndpointSlice) {
+	newAddressSet := pp.endpointSliceToAddresses(slice)
+	for id, addr := range pp.addresses.Addresses {
+		newAddressSet.Addresses[id] = addr
+	}
+
+	addAddress := make(map[ID]Address, len(newAddressSet.Addresses))
+	for id, newAddress := range newAddressSet.Addresses {
+		if _, ok := pp.addresses.Addresses[id]; !ok {
+			addAddress[id] = newAddress
+		}
+	}
+
+	if len(addAddress) > 0 {
+		add := AddressSet{
+			Addresses: addAddress,
+			Labels:    newAddressSet.Labels,
+		}
+		for _, listener := range pp.listeners {
+			listener.Add(add)
+		}
+	}
+
 	pp.addresses = newAddressSet
+	pp.exists = true
+	pp.metrics.incUpdates()
+	pp.metrics.setPods(len(pp.addresses.Addresses))
+	pp.metrics.setExists(true)
+}
+
+func (pp *portPublisher) updateEndpointSlice(oldSlice *discovery.EndpointSlice, newSlice *discovery.EndpointSlice) {
+	newAddressSet, oldAddressSet := pp.endpointSliceToAddresses(newSlice), pp.endpointSliceToAddresses(oldSlice)
+	if len(newAddressSet.Addresses) == 0 && len(oldAddressSet.Addresses) == 0 {
+		return
+	}
+
+	for id, addr := range pp.addresses.Addresses { // what happens when len = 0? we can add new addresses but we can't remove any
+		// we want to remove old slice addresses =>
+		// if we have any, don't add 'em
+		if _, found := oldAddressSet.Addresses[id]; found {
+			continue
+		}
+
+		newAddressSet.Addresses[id] = addr
+	}
 
+	add, remove := diffAddresses(pp.addresses, newAddressSet)
+	for _, listener := range pp.listeners {
+		if len(remove.Addresses) > 0 {
+			listener.Remove(remove)
+		}
+		if len(add.Addresses) > 0 {
+			listener.Add(add)
+		}
+	}
+
+	pp.addresses = newAddressSet
+	if len(pp.addresses.Addresses) == 0 {
```

Comment:
If after updating we end up with no addresses (1 service, 1 slice, oldSlice = 1 IP address, newSlice = nil ports, nil endpoints, for example) we should send a `NoEndpoints`. So far this seems to work well, looking for some feedback here if this makes sense logically?

Reply:
In the Endpoints code, if we get an update with 0 addresses, we know that all addresses have been deleted so we send a NoEndpoints which I think is effectively a shorthand for "remove all".  In the the case of EndpointSlices the logic is more complicated because an updated slice with 0 addresses doesn't mean that there are no addresses for the whole service.  I think we can do ourselves a favor here and just not use the `NoEndpoints` shortcut at all and just rely on plain Adds and Removes for managing things.

Reply:
Oh! This makes so much sense then. I've been under the impression constantly when testing that when all addresses are removed the `no endpoints` message should be sent. This is why I was so focused on putting it in there. Thank you for clarifying, it's more intuitive just doing it with adds and removes I think so will remove it (the no endpoints call).